### PR TITLE
Issue 6811: LTS - Always disable lazy commit functionality for all writes to all segments

### LINKED ISSUE #6811: Add integrity checks in Pravega (LTS)
**Is your feature request related to a problem? Please describe.**
We need a mechanism to check the integrity of data just written to LTS. Clearly, checking data integrity has a performance impact. For this reason, we may want to use this feature mainly during the development and testing cycles of Pravega. It could be used also in production, but considering the potential impact on performance.

**Describe the solution you'd like**
Check integrity of data written to LTS:
  - Once we write data to LTS, we need to make sure that what has been written is correct. This may entail re-reading the written bytes.
  - Note that doing so when reading data from LTS may be more complex, given that we may read arbitrary offsets of chunks in storage.

Note that the approach taken here may be more biased towards ensuring correctness during development and testing cycles, to proactively detect and correct any potential data corruption issues that could be related to Pravega. 

**Describe alternatives you've considered**
n/a

**Additional context**
This feature may be very useful to discover data corruption issues proactively during development and testing to avoid them to happen in production.

### PR DESCRIPTION
**Pull request by sachin-j-joshi**
**Change log description**  
Always disable lazy commit functionality for all writes to all segments #6841
Make non-lazy commit mode default for LTS.
Support seamless migration of deployment with lazy commit enabled to one with lazy commit disabled.

**Purpose of the change**  
Fixes #6811

**What the code does**  
- Make non-lazy commit mode default for LTS.
- With this change all attribute segments are always written atomically.
- Track atomicity via flag on segment metadata 
- Add `AppendRecord` to `SystemJournal `after each append to system segments
- Add ability to correctly migrate from lazy commit deployment to non-lazy commit deployment.
- Fix tests as necessary

**How to verify it**  
All tests should pass.

### CODE REVIEW

**Comment by RaulGracia** [comment]
@sachin-j-joshi I have done a manual test with this branch, and I still can reproduce the problem with BTreeIndex reading a corrupted footer if I manually add few bytes to an Attribute Index Segment:
- Build and deploy locally this branch, using File System as LTS.
- Run some sample or application to generate some data and see that it works.
- Pick an Attribute Index Segment and add some bytes to it (e.g., Scopes Table Segment, for instance)
- Restart the Segment Store process.

My expectation is that, with this change, the BTreeIndex would be getting the offset of the last successful write via `storage.getSegmentInfo()`. However, we can see the error during the initialization of BTreeIndex.

**Comment by sachin-j-joshi** [comment]
I'm not sure how this could be possible. There is no code path that adjusts metadata length of last chunk except inside claimOwnership which is now turned off by default.

**Comment by sachin-j-joshi** [comment]
Also I think to just avoid any future confusion let's always turn off optimization on system segments and any attribute segments irrespective of whether lazy commit set or not.

**Comment by RaulGracia** [comment]
@sachin-j-joshi if that optimization induces violation of atomicity, perhaps it shouldn't be allowed at all on any segment and should be just removed to simplify the codebase, right? Maybe we can find other places to optimize metadata updates while guaranteeing atomicity (I remember that `WriterTableProcessor`, `IndexWriter` and/or `AttributeAggregator` may already contain optimizations when there are multiple updates on the same key for a given batch to process, as happens for frequent updates of the `lastChunk` length).

But even before that: before attempting to optimize anything, I still miss a quantitative measurement of the metadata updates that happen with and without `lazyCommit`. Because without numbers, `lazyCommit` is perhaps trying to reduce some cost in metadata updates that is actually negligible at the expense of losing atomicity in SLTS. Please, could you please do an analysis of this (e.g., number of Storage Metadata Table Segment updates per segment w/wo `lazyCommit`)? How does it impact in terms of LTS operations? Without this data (that can be observed locally), we are just guessing to keep some optimization code that is actually problematic and may not be even needed.

**Comment by RaulGracia** [comment]
@sachin-j-joshi I have performed again the test scenario mentioned above, and now it works. I think that the problem was with the test approach when editing the attribute index file (the editor may have changed some characters from the binary file). In the new test I just edited the file by appending data like `echo hola >> 'scopes$attributes.index.E-1-O-0.42e3f5da-6e99-48e9-b2eb-a46eead3115c'`, which keeps the rest of contents of the file untouched.

In the new reproduction, I can confirm that the `BTreeIndex` for scopes has been able to be properly initialized few times after appending garbage data to it, whereas without this PR I'm able to get the `DataCorruptionException` doing the same test.

**Comment by sachin-j-joshi** [comment]
@Aditi Jadhav @amit.singh30 Please review this change.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java [hunk 1]
```diff
@@ -308,10 +308,15 @@ public static ChunkedSegmentStorage getReference(Storage storage) {
      *                        throws StorageMetadataException In case of any chunk metadata store related errors.
      */
     private CompletableFuture<Void> claimOwnership(MetadataTransaction txn, SegmentMetadata segmentMetadata) {
+        Preconditions.checkState(!segmentMetadata.isStorageSystemSegment(), "claimOwnership called on system segment %s", segmentMetadata);
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java:311]: Question: the reason why we still keep all the reconciliation stuff here is to make this change backwards compatible, right? Because we may be upgrading from a Pravega cluster with `lazyCommit=true` to a a cluster with this change, in which case we need to also reconcile as the data/metadata left in the system may have not been atomically updated, right? If so, why system segments are not reconciled?

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java [hunk 2]
```diff
@@ -385,6 +391,10 @@ private CompletableFuture<Void> claimOwnership(MetadataTransaction txn, SegmentM
         }, executor);
     }
 
+    boolean isAlwaysAtomic(SegmentMetadata segmentMetadata) {
+        return NameUtils.isAttributeSegment(segmentMetadata.getName());
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java:395]: Are we sure that `lazyCommit` was disabled for attribute segments and system segments?

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 3]
```diff
@@ -1602,6 +1641,76 @@ private void read00(RevisionDataInput input, TruncationRecord.TruncationRecordBu
         }
     }
 
+    /**
+     * Journal record for segment append.
+     */
+    @Builder(toBuilder = true)
+    @Data
+    @EqualsAndHashCode(callSuper = true)
+    static class AppendRecord extends SystemJournalRecord {
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1650]: If there is a rollback when upgrading to this PR and some old Segment Store finds a system journal with an `AppendRecord`, would it be skipped or will make the initialization process to fail? can we add a test for that?

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StatusFlags.java [hunk 4]
```diff
@@ -33,6 +33,11 @@ public final class StatusFlags {
      */
     public static final int SEALED = 0x0002;
 
+    /**
+     * Flag to indicate whether all writes are atomic.
+     */
+    public static final int ATOMIC_WRITES = 0x0004;
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StatusFlags.java:39]: Is this something encoded in the Storage Metadata information of each Segment? Could this break anything or is it using some unused space to encode more information for a Segment? Why using `0x0004` instead of `0x0003`, which looks like the next number to choose?

**Review by RaulGracia — CHANGES_REQUESTED** [review]
Looks good, I have some comments. Overall, my main concerns are that this change i) does not break anything in terms of compatibility, and ii) that performance is not significantly degraded. Also, we need to run system tests and longevity with this change.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java (hunk 1 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java:311]: 1. Yes , we have kept the reconciliation logic to handle upgrade.
2. System segments are always recovered as first step during SLTS bootstrap via replaying of journal entries. Because they are reconciled every time container restarts , `claimOwnership` is never actually called on them (the metadata is already updated during bootstrap.)
3. SystemJournal logic takes care of migration scenarios and unit tests cover these scenarios separately

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java (hunk 2 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java:395]: This code is checking whether given segment is supposed to be always atomic ? Answer is yes if it is attribute segment

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java [hunk 5]
```diff
@@ -409,6 +419,7 @@ public CompletableFuture<SegmentHandle> create(String streamSegmentName, Segment
                                     .build();
 
                             newSegmentMetadata.setActive(true);
+                            newSegmentMetadata.setAtomicWrites(isAlwaysAtomic(newSegmentMetadata) || !config.isLazyCommitEnabled());
```
[the remark this answers is not in the document]
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java:422]: persisted in storage metadata

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 6]
```diff
@@ -953,6 +949,12 @@ private CompletableFuture<Void> applyRecord(MetadataTransaction txn,
                     nullToEmpty(chunkAddedRecord.getOldChunkName()),
                     chunkAddedRecord.getNewChunkName(),
                     chunkAddedRecord.getOffset());
+        } else if (record instanceof AppendRecord) {
+            val appendRecord = (AppendRecord) record;
```
[the remark this answers is not in the document]
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:953]: yes. Here we are removing performance optimization.
I expect the perf overhead to be minimal but haven't yet run perf bench marks.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 7]
```diff
@@ -1002,9 +1004,10 @@ private CompletableFuture<Void> adjustLastChunkLengths(MetadataTransaction txn)
                             ff = CompletableFuture.completedFuture(null);
                         }
                         return ff.thenApplyAsync(v -> {
+                            // Always set as atomic write
+                            segmentMetadata.setAtomicWrites(true);
```
[the remark this answers is not in the document]
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1008]: With new version we always want all system segments to be atomic.
Note that this code is recovering state of system segments from journals. Here at the end of that recovery we are saying that going forward system segments need to be updated atomically. hence setting flag to be true

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StatusFlags.java (hunk 4 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StatusFlags.java:39]: It is a bit flag 
first bit -> is active or not   00001
second bit -> is sealed or not  00010
third bit -> is atomic or not  00100
fourth -> has ownership changed  01000 
fourth -> is system segment  10000 


we have 32 flags totals - only 5 used so far

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StatusFlags.java:39]: The status flag field is an integer and we use it as bitarray

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java (hunk 3 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1650]: good point. 
let me know how best to handle the roll back.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 8]
```diff
@@ -1111,6 +1114,41 @@ private CompletableFuture<Void> applyChunkAddition(MetadataTransaction txn, Map<
...
+                            .thenAcceptAsync( metadata -> {
+                                Preconditions.checkState(metadata != null, "metadata must not be null");
+                                ChunkMetadata lastChunkMetadata = (ChunkMetadata) metadata;
+                                Preconditions.checkState(lastChunkMetadata.getName().equals(chunkName), "Invalid chunk name expected= %s actual=%s in segment=%s",
+                                        lastChunkMetadata.getName(), chunkName, segmentMetadata);
+                                Preconditions.checkState(lastChunkMetadata.getLength() == offset, "Invalid offset expected= %s actual=%s in segment=%s",
+                                        lastChunkMetadata.getLength(), offset, segmentMetadata);
+                                segmentMetadata.setLength(segmentMetadata.getLength() + length);
+                                lastChunkMetadata.setLength(lastChunkMetadata.getLength() + length);
+                                segmentMetadata.setAtomicWrites(true);
```
Comment (abhinb) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1141]: does setAtomicWrites need to be set everytime we update the journal?

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1141]: Segment metadata for all segment is stored in metadata tables. Once marked in metadata we do not need to mark it again.
1. For all new segments we mark appropriately during create operation
2. For old segments we update that flag when we open that segment for the first time after new code and update the metadata 
3. For system segments, the system snapshot the metadata is in system journals. We recreate in-memory metadata for system segments by replaying journals during boot. This particular line is called on bootstrap and always mark system segments as atomic

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 9]
```diff
@@ -1111,6 +1114,41 @@ private CompletableFuture<Void> applyChunkAddition(MetadataTransaction txn, Map<
                 }, executor);
     }
 
+    /**
+     * Apply chunk append.
```
[the remark this answers is not in the document]
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1118]: yes we only added "chunk added" record and did length reconciliation only after failover.
As you know this approach is not atomic. Therefore now we always add append to journal to be always atomic.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java [hunk 10]
```diff
@@ -1602,6 +1641,76 @@ private void read00(RevisionDataInput input, TruncationRecord.TruncationRecordBu
         }
     }
 
+    /**
+     * Journal record for segment append.
```
Comment (tkaitchuck) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1645]: I don't understand how this ties into what is described in the PR. If we're just removing an optimization, which was only sometimes on, why is there a new journal record type?

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java [hunk 11]
```diff
@@ -172,6 +172,16 @@ public SegmentMetadata setOwnershipChanged(boolean value) {
         return setFlag(StatusFlags.OWNERSHIP_CHANGED, value);
     }
 
+    /**
+     * Sets whether writes should be atomic.
+     * @param value Value to set.
+     * @return This instance so that these calls can be chained.
+     */
+    public SegmentMetadata setAtomicWrites(boolean value) {
+        return setFlag(StatusFlags.ATOMIC_WRITES, value);
```
Comment (tkaitchuck) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java:181]: Based on my conversation with @RaulGracia and this comment: https://github.com/pravega/pravega/pull/6827#discussion_r975394542
I am not sure this is something we ever want off. Because it would seem that while we gain a minor performance gain in the steady state, the other PR we would need to add would cost us quite a bit during failover.

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java:181]: Atomicity of write operation in SLTs comes with the fine print and an exception.
As described [here](https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MetadataTransaction.java#L55-L57) Lazy commit functionality requires reconciliation. That is a clearly expected assumption and stated assumption for turning it on.

**How it works**
- With this optimization turned on, SLTS will not commit metadata changes if only change to metadata is an update to the length of segment + existing last chunk during append. In such cases updated metadata is kept in memory. 
- In all other cases metadata is always committed to table segment and up to date.
- In normal operation SLTS can guarantee atomic writes in all cases. Either whole data is appended and visible or none.
- In case of failure over, updated metadata in memory is lost and metadata in table segment is stale. So we need to reconcile and adjust length of last chunk, which happens in [claimOnwership](https://github.com/pravega/pravega/blob/01903b9ee88c8845a242b37bc565f2b1b7b32ca5/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java#L310) . We simply look at the current length of last chunk on LTS and adjust its metadata.  This reconciliation happens during [first openWrite after failure](https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java#L255-L258)
- [Similar logic](https://github.com/pravega/pravega/blob/01903b9ee88c8845a242b37bc565f2b1b7b32ca5/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java#L975) is implement while re-playing journal entries and recreating state of storage and container metadata segments.  
- A strong assumption in this recovery procedure is that upper layers are able to deals with partially written data and smartly handle it.
- This  works wonderfully for normal segments.

**What doesn't work**
In case of fail over SLTS breaks atomic update behavior ..in this one odd case you'll find partial data on LTS and that data will be reconciled.

**What is the problem?**

- For the reasons I am not totally convinced about the table segment BTree index logic makes an incorrect assumption that   SLTS writes are always atomic - so either whole page is written or none at all. Even in cases of failure it assumes that page was written entirely. It blindly assumes that last 12 bytes of segment always contain header.
- The big issue here is that btree index segment does not have any reconciliation logic whatsoever and it does not store any state in BK and can not recreate same page with same data again. 
- This is the real problem. Btree index logic needs to be fixed. Period.

**Where should we fix it?**
1. We can fix the atomicity in SLTS 
2. Or fix it in BTree by implementing reconciliation logic
My preference is that we fix the Btree logic and not mess with SLTS optimization

**Why new type of journal entry**
- With lazy commit turned off, the SLTS journal needs a record about each append operation  
- Originally we assumed that upper layers can handle partial writes , so never cared to make write operation 100% atomic even after failure. 
- Seems like if we don't want to fix Btree then we are left with no choice and are forced to implement atomicity in SLTS 100% of times.


I hope that answers your question.
My first preference is to fix Btree code instead of this  change.

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java:181]: Note that this is an optimization that is relevant only in append mode. In no append mode it is always turned off. And because chunks are immutable and we add new chunk for each write in no-append mode, we don't have to deal with partial writes. 
In case of fail-over, [chunk creation] https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorage.java#L112-L123 will never succeed in the first place

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java (hunk 10 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1645]: see below...

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java (hunk 11 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java:181]: @tkaitchuck Another way to answer your question is to look at this flag as way to handle migration from data that was written by lazy-commit mode to code that has lazy commit mode turned off.

Note that currently lazy commit is on by default, So all data today is written assuming that after restart the new instance will check the length and 

When lazy commit is turned off we always commit metadata and there is no reason reconcile length of metadata based on length of LTS. (because in SLTS metadata was updated only after write was written successfully and fully)

So this flag is a marker flag in two sense
- when this flag is off then it indicates the data was written expecting length reconciliation. Any data written this change needs reconciliation (aka length adjustment) to work correctly. Therefore this change will reconcile length for any data written before this version
- when this flag is on then lazy commit can be never used with this segment again. And indicates that no length reconciliation is needed  any more.

This flag helps us migrate correctly from old version to new.

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java:181]: Remove the option to configure lazy commit. From now on all segments are always written atomically.

**Review by tkaitchuck — APPROVED** [review]

**Comment by RaulGracia** [comment]
@sachin-j-joshi, please could you summarize what is the procedure in case of a rollback of a version of Pravega with this version and an older version of Pravega without it? Should we create an issue to add the journal serializer to previous release versions? Even if we do that, is there any guarantee that the rollback would succeed?

**Comment by sachin-j-joshi** [comment]
https://github.com/pravega/pravega/pull/6909 For porting changes to 0.12 branch

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java (hunk 3 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java:1655]: https://github.com/pravega/pravega/pull/6909

**Review by RaulGracia — APPROVED** [review]
After latest changes, it seems that this PR works. LGTM.
